cmd/syncComparator.go

```diff
@@ -54,30 +202,217 @@ func newSyncDestinationComparator(i *folderIndexer, copyScheduler, cleaner objec
+			//    newly created in the source since last sync. We need to copy all of them to the target.
+			//
+			lcFolderName = path.Join(lcFolderName, lcFileName)
+			foldermap, folderPresent = f.sourceFolderIndex.folderMap[lcFolderName]
+			if folderPresent {
+				fmt.Printf("Finalizing directory %s (FinalizeAll=%v)\n", lcFolderName, destinationObject.isFinalizeAll)
+				f.FinalizeTargetDirectory(foldermap, destinationObject.isFinalizeAll)
+			}
+			return nil
+		} else {
+			if folderPresent {
+				sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+				// Lets check if entry in source is folder and entry on destination should be folder.
+				// Otherwise folder in source changed to file.
+				if present && (destinationObject.entityType == sourceObjectInMap.entityType) {
+					delete(foldermap.indexMap, lcFileName)
+					size := storedObjectSize(sourceObjectInMap)
+					size = -size
+					atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
+					if sourceObjectInMap.relativePath != destinationObject.relativePath {
+						panic("Relative Path not matched")
+					}
+					delete(f.sourceFolderIndex.folderMap[lcFolderName].indexMap, lcFileName)
+
+					// In case of folder if ctime/mtime changed, lets create the folder.
+					dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+					if f.disableComparison || dataChanged || metadataChanged {
+						err := f.copyTransferScheduler(sourceObjectInMap)
+						if err != nil {
+							return err
+						}
+					}
+					return nil
+				}
+			}
+			// We detect folder not present on source, now we need to delete the folder and files underneath.
+			// Other case will be folder at source changed to file, so we need to delete the folder in target.
+			// TODO: Need to add call to delete the folder.
+			_ = f.destinationCleaner(destinationObject)
+
+			return nil
+		}
 	}
 
+	// File case.
+	// Parent Folder present, we need to check if file exists or not.
+	if folderPresent {
+		sourceObjectInMap, present = foldermap.indexMap[lcFileName]
+
+		// if the destinationObject is present at source and stale, we transfer the up-to-date version from source
+		if present && (sourceObjectInMap.entityType == destinationObject.entityType) {
+			// Sanity check.
+			if sourceObjectInMap.relativePath != destinationObject.relativePath {
+				panic("Relative Path not matched")
+			}
+
+			dataChanged, metadataChanged := f.HasFileChangedSinceLastSyncUsingTargetCompare(destinationObject, sourceObjectInMap)
+			if f.disableComparison || dataChanged {
+				err := f.copyTransferScheduler(sourceObjectInMap)
+				if err != nil {
+					return err
+				}
+			} else if metadataChanged {
+				// TODO: Need to add call to just update the metadata only.
+			}
+			size := storedObjectSize(sourceObjectInMap)
+			size = -size
+			atomic.AddInt64(&f.sourceFolderIndex.totalSize, size)
```

Comment:
need to be careful we don't miss any of these.. I prefer adding panic in all unsupported places so that we know if it runs this code.
also, for uniformity use

atomic.AddInt64(&f.sourceFolderIndex.totalSize, -(storedObjectSize(sourceObjectInMap)))

in all places